modules/aruco/src/aruco.cpp

```diff
@@ -244,39 +247,86 @@ static void _filterTooCloseCandidates(const vector< vector< Point2f > > &candida
+                        vector<unsigned int> grouped;
+                        grouped.push_back(i);
+                        grouped.push_back(j);
+                        groupedCandidates.push_back( grouped );
+                    }
+                    // i is related to a group
+                    else if(candGroup[i] > -1 && candGroup[j] == -1){
+                        int group = candGroup[i];
+                        candGroup[j] = group;
+
+                        // add to group
+                        groupedCandidates[group].push_back( j );
+                    }
+                    // j is related to a group
+                    else if(candGroup[j] > -1 && candGroup[i] == -1){
+                        int group = candGroup[j];
+                        candGroup[i] = group;
+
+                        // add to group
+                        groupedCandidates[group].push_back( i );
+                    }
                 }
             }
         }
     }
 
-    // mark smaller one in pairs to remove
-    vector< bool > toRemove(candidatesIn.size(), false);
-    for(unsigned int i = 0; i < nearCandidates.size(); i++) {
-        // if one of the marker has been already markerd to removed, dont need to do anything
-        if(toRemove[nearCandidates[i].first] || toRemove[nearCandidates[i].second]) continue;
-        size_t perimeter1 = contoursIn[nearCandidates[i].first].size();
-        size_t perimeter2 = contoursIn[nearCandidates[i].second].size();
-        if(perimeter1 > perimeter2)
-            toRemove[nearCandidates[i].second] = true;
-        else
-            toRemove[nearCandidates[i].first] = true;
-    }
+    // save possible candidates
+    candidatesSetOut.clear();
+    contoursSetOut.clear();
+
+    vector< vector< Point2f > > biggerCandidates;
+    vector< vector< Point > > biggerContours;
+    vector< vector< Point2f > > smallerCandidates;
+    vector< vector< Point > > smallerContours;
+
+    // save possible candidates
+    for( unsigned int i = 0; i < groupedCandidates.size(); i++ ) {
+        int smallerIdx = groupedCandidates[i][0];
+        int biggerIdx = -1;
+
+        // evaluate group elements
+        for( unsigned int j = 1; j < groupedCandidates[i].size(); j++ ) {
+            size_t currPerim = contoursIn[ groupedCandidates[i][j] ].size();
+
+            // check if current contour is bigger
+            if ( biggerIdx < 0 )
+                biggerIdx = groupedCandidates[i][j];
+            else if(currPerim >= contoursIn[ biggerIdx ].size())
+                biggerIdx = groupedCandidates[i][j];
+
+            // check if current contour is smaller
+            if(currPerim < contoursIn[ smallerIdx ].size() && detectInvertedMarker)
+                smallerIdx = groupedCandidates[i][j];
+        }
+        // add contours und candidates
+        if(biggerIdx > -1){
+
+            biggerCandidates.push_back(candidatesIn[biggerIdx]);
+            biggerContours.push_back(contoursIn[biggerIdx]);
```

Comment:
if we observing lack of performance in this code, then we can try to utilize smart pointers:

```.diff
-const vector< vector< Point2f > > &candidatesIn
+typedef vector< Point2f > CandidatePoints;
+typedef cv::Ptr<CandidatePoints> CandidatePointsPtr;
+const vector<CandidatePointsPtr> &candidatesIn
```